ignition: run files stage at original CL ordering

jlebon opened the pull request:
We previously messed around with changing the files service to
adapt to `ostree-prepare-root.service` running so late in the boot
process. Now that OSTree runs at more semantically appropriate times, we
can restore ordering here so that it runs right after
`initrd-root-fs.target`.

This also undoes the hacking around in `ignition-generator` to get the
services pulled in by `initrd-switch-root.service` instead of the more
appropriate principal target of the initrd (`initrd.target`).

Another nice thing about this is that we drop the explicit dependency
between `ostree-prepare-root.service` and `ignition-files` since they're
now synchronized through `initrd-root-fs.target`.

Closes: #20

Requires: https://github.com/ostreedev/ostree/pull/1759

jlebon commented:
Split out the first two commits in #27.

jlebon commented:
/cc @lucab

jlebon commented:
More generally, this is prep for making it easier to support `/var` mountpoints (see #18).

dustymabe commented:
For the first commit: 

```
remove ignition-remount-sysroot.service files

These are no longer needed now that we explicitly pass `rw` on the
kernel command-line. For more information, see:
```

We should probably start some docs for requirements. rw on the kernel command line being required can be the first one.

dustymabe commented:
```
(Anyway, I think we want `network-online.target` here instead of
`network.target` but I haven't tested that).
```

yes. I believe from my reading the recommendation is that `network-online.target` is what people generally want when they specify `network.target`. I agree. I think when I was looking into this in the past `network-online.target` didn't exist for el7, so might be something we look at before we make a change.

dustymabe commented:
everything LGTM. I'll defer to @lucab though

dustymabe commented:
Also I really like the fact that we now don't have anything in here specific for ostree based systems, which means it's easier for other OS to pick this up and use it if they want.

#27 merged so you can rebase this now

jlebon commented:
Rebased! (But note this needs to wait until https://github.com/ostreedev/ostree/pull/1759 is merged, I changed the title).

> Also I really like the fact that we now don't have anything in here specific for ostree based systems

Yup! The only thing left is the `ask-var-mount` stuff which will be dropped once we fix #18.

lucab commented:
The first chunk of that was already LGTM from my side, and this one is also aligned with I originally reported. I'll try to allocate some time next week to slowly go through the deps-graph next week. @jlebon if you are testing this, can you please do this with an Ignition file that does both FS mounting and file writing?

jlebon commented:
> @jlebon if you are testing this, can you please do this with an Ignition file that does both FS mounting and file writing?

Yup, will do!

jlebon commented:
So one thing related to this that I've just noticed is that F28 traditional cloud images ship with `ro` on the kernel cmdline. So we'd have to change that there if we want to eventually use Ignition there too. (Not just for Ignition, but also for presets and `ConditionFirstBoot=`). I don't think this should be a blocker for dropping the compat `remount-sysroot` units though.

dustymabe commented:
> So we'd have to change that there if we want to eventually use Ignition there too.

i'd be willing to make that change. I think if `rw` is a requirement then we should have some error checking and nice messages to users to tell them when we detect `ro`. this will certainly help **us** when debugging issues users report. 

@jlebon WDYT? do you know of a good way to achieve this?

jlebon commented:
OK, I made a build of systemd for F29 with https://github.com/systemd/systemd/pull/9625 backported here: https://jlebon.fedorapeople.org/systemd-ign/.

You can use the [overrides/rpm](https://github.com/coreos/coreos-assembler/pull/145) dir or use this one-liner if you've already got a system setup the way you want:

```
rpm-ostree override replace https://jlebon.fedorapeople.org/systemd-ign/systemd-{,{pam,libs,udev}-}239-4.jl.fc29.x86_64.rpm
```

With that systemd + this PR + https://github.com/ostreedev/ostree/pull/1759, I can successfully provision a system with a secondary disk on FCOS using a config more or less similar to @lucab in https://github.com/dustymabe/ignition-dracut/pull/17#issuecomment-422334360.

I can confirm the order is still: `ignition-disks.service`, `ostree-prepare-root.service`, `initrd-root-fs.target`, and then `ignition-files.service`. I was trying to use `systemd-analyze plot` to show this in a nicer way, but it doesn't know about initrd services (and trying to run it from the initrd just gives "boot not completed").

jlebon commented:
> @jlebon WDYT? do you know of a good way to achieve this?

Well, the read-only error message will be pretty clear :). If we really want an explicit error upfront, we could probably have a unit like:

```
[Unit]
Description=Check for Writable Sysroot
DefaultDependencies=no
ConditionPathIsReadWrite=|!/sysroot
ConditionPathIsReadWrite=|!/sysroot/var
After=initrd-root-fs.target
OnFailure=emergency.target

[Service]
Type=oneshot
ExecStart=/usr/bin/echo "Sysroot is not mounted rw!"
ExecStart=/usr/bin/false
```

?

We could maybe use `ConditionKernelCommandLine=!rw` instead though it's not as generic.

dustymabe commented:
This looks pretty good. Maybe we should even give them a little more:

- *"Sysroot is not mounted rw! For Ignition dracut modules please make sure rw (not ro) is provided on kernel command line."*

jlebon commented:
OK, I dropped the `[DONOTMERGE]` tag.

One thing I just realized actually is that not only does this require https://github.com/ostreedev/ostree/pull/1759, but they're both dependent on each other. (So systems with that OSTree patch but without this one will not boot). So, this will break RHCOS streams until this one gets merged. Hmm, probably easier to just freeze OSTree before that merge.

jlebon commented:
https://github.com/CentOS/sig-atomic-buildscripts/pull/359

lucab commented on dracut/30ignition/ignition-ask-var-mount.service:
```diff
@@ -7,7 +7,6 @@
 [Unit]
 Description=Ask OSTree to mount /var in initramfs
 DefaultDependencies=false
-After=initrd-switch-root.target
```
I don't see any immediate problems with this, but I'd be wary of leaving this unit completely without a lower bound. I'd suggest to make this a `After=basic.target`.

lucab commented on dracut/30ignition/ignition-files.service:
```diff
@@ -5,30 +5,14 @@ DefaultDependencies=false
+Before=initrd-parse-etc.service
+
+Requires=ignition-disks.service
+After=ignition-disks.service
```
This should be already implicit in the `After=initrd-root-fs.target` above. We can leave this explicit for the sake of clarity if we want. Maybe we can merge these with the lines above though, as the comment there already explains it.

lucab commented on dracut/30ignition/ignition-files.service:
```diff
@@ -5,30 +5,14 @@ DefaultDependencies=false
 # Ignition files stage starts after partitioning is completed.
 Requires=initrd-root-fs.target
 After=initrd-root-fs.target
+Before=initrd-parse-etc.service
```
Would you mind leaving a comment above here to remind that Ignition can setup systemd config in the real rootfs, which are picked up by this service, thus the ordering?

lucab commented on dracut/30ignition/ignition-files.service:
```diff
@@ -5,30 +5,14 @@ DefaultDependencies=false
-Requires=ignition-remount-sysroot-var.service
-After=ignition-remount-sysroot-var.service
-
-After=ostree-prepare-root.service
```
I think this line is still relevant and should be kept somewhere else, not dropped.

In particular, I think that with this PR `ignition-disks` and `ostree-prepare-root` are now racing with each other. This may be a problem when ignition tries to format rootfs or /var.

I think a `Before=ostree-prepare-root.service` in ignition-disks would fix this, but please double check me.

jlebon commented:
OK, I pushed a `fixup!` to make it easier for folks to see the diff. I can squash it in once we're approved since we don't have Homu here.

> In particular, I think that with this PR `ignition-disks` and `ostree-prepare-root` are now racing with each other.

Hmm yeah, I think you're right. So, looking at the original [upstream version](https://github.com/coreos/bootengine/blob/760986ba0aa3601e60ab0057ea34370d496b7dfc/dracut/30ignition/ignition-disks.service) of `ignition-disks.service`, it used to be:

```
Requires=local-fs-pre.target
Before=local-fs-pre.target
```

Though in this fork, that was changed in #1 to:

```
Requires=sysinit.target
After=sysinit.target
```

I changed it to instead be `Before=sysroot.mount` with a comment to explain.

jlebon commented:
OK, retested on FCOS with that fixup! :heavy_check_mark:

lucab commented on dracut/30ignition/ignition-disks.service:
```diff
@@ -4,8 +4,13 @@ DefaultDependencies=false
 # see https://www.freedesktop.org/software/systemd/man/bootup.html
+
+# Note that CL runs this before `local-fs-pre.target` to allow for configs that
+# completely wipe the rootfs. Though we're not there yet. But we still run #
```
Spurious trailing `#`.

jlebon commented:
OK, now squashed! :arrow_up:

jlebon commented:
Once this is merged, we need to:

1. undo https://github.com/CentOS/sig-atomic-buildscripts/pull/359
2. bump `ignition-dracut` in Fedora as soon as https://github.com/ostreedev/ostree/pull/1761 is released and built so that FCOS doesn't break

dustymabe commented:
:)

ok merging. We'll coordinate the copr build with the two steps in https://github.com/dustymabe/ignition-dracut/pull/26#issuecomment-432748722
